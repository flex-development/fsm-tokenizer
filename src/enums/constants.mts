/**
 * @file Enums - constants
 * @module vfile-tokenizer/enums/constants
 */

/**
 * Constant values.
 *
 * @enum {number}
 */
const enum constants {
  tabSize = 2,
  v8MaxSafeChunkSize = 10_000
}

export default constants
