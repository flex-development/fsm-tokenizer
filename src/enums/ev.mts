/**
 * @file Enums - ev
 * @module fsm-tokenizer/enums/ev
 */

import type { EventType } from '@flex-development/fsm-tokenizer'

/**
 * Event types.
 *
 * @enum {EventType}
 */
enum ev {
  enter = 'enter',
  exit = 'exit'
}

export default ev
