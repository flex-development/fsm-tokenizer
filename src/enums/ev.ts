/**
 * @file Enums - ev
 * @module vfile-lexer/enums/ev
 */

/**
 * Event types.
 *
 * @enum {Lowercase<string>}
 */
enum ev {
  enter = 'enter',
  exit = 'exit'
}

export default ev
