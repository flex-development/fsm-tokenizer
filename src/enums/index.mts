/**
 * @file Entry Point - Enums
 * @module vfile-tokenizer/enums
 */

export { default as chars } from '#enums/chars'
export { default as codes } from '#enums/codes'
export { default as ev } from '#enums/ev'
