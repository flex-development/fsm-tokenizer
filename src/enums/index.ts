/**
 * @file Entry Point - Enums
 * @module vfile-lexer/enums
 */

export { default as ev } from './ev'
export { default as tt } from './tt'
