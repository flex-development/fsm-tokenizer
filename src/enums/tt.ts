/**
 * @file Enums - tt
 * @module vfile-lexer/enums/tt
 */

/**
 * Token types.
 *
 * @enum {Lowercase<string>}
 */
enum tt {
  eof = 'eof',
  sof = 'sof'
}

export default tt
