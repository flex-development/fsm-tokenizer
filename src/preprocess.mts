/**
 * @file preprocess
 * @module vfile-tokenizer/preprocess
 */

import codes from '#enums/codes'
import constants from '#enums/constants'
import type {
  Code,
  Column,
  Encoding,
  FileLike,
  Preprocess,
  PreprocessOptions,
  Value
} from '@flex-development/vfile-tokenizer'

/**
 * Create a preprocessor to turn a value into character code chunks.
 *
 * @see {@linkcode Preprocess}
 * @see {@linkcode PreprocessOptions}
 *
 * @this {void}
 *
 * @param {PreprocessOptions | null | undefined} [options]
 *  Configuration options
 * @return {Preprocess}
 *  Character code preprocessor
 */
function preprocess(
  this: void,
  options?: PreprocessOptions | null | undefined
): Preprocess {
  return Object.defineProperties(preprocessor.bind(options ?? {}), {
    name: { value: 'preprocess' }
  })

  /**
   * Turn `value` into character code chunks.
   *
   * @this {PreprocessOptions}
   *
   * @param {FileLike | Value | null | undefined} value
   *  The value to preprocess
   * @param {Encoding | null | undefined} [encoding]
   *  The character encoding to use when `value`
   *  or its contents is {@linkcode Uint8Array}
   * @param {boolean | null | undefined} [end]
   *  Whether the end of stream has been reached
   * @return {Code[]}
   *  Character code chunks
   */
  function preprocessor(
    this: PreprocessOptions,
    value: FileLike | Value | null | undefined,
    encoding?: Encoding | null | undefined,
    end?: boolean | null | undefined
  ): Code[] {
    /**
     * Character code chunks.
     *
     * @const {Code[]} chunks
     */
    const chunks: Code[] = []

    /**
     * Number of spaces a tab is equivalent to.
     *
     * @const {number} tabSize
     */
    const tabSize: number = this.tabSize ?? constants.tabSize

    if (value === null) {
      chunks.push(codes.break)
    } else if (value !== undefined) {
      if (typeof value === 'object' && 'value' in value) {
        value = value.value
      }

      if (typeof value !== 'string') {
        value = new TextDecoder(encoding ?? undefined).decode(value)
      }

      /**
       * Current column.
       *
       * @var {Column} column
       */
      let column: Column = 1

      /**
       * Index of current character code.
       *
       * @var {number} index
       */
      let index: number = 0

      while (index < value.length) {
        /**
         * Character code.
         *
         * @var {NonNullable<Code>} code
         */
        let code: NonNullable<Code> = value[index]!.codePointAt(0)!

        /**
         * Difference between next column and current column.
         *
         * @var {number} k
         */
        let k: number = 1

        switch (true) {
          case code === codes.cr:
            if (value[index + 1]?.codePointAt(0) === codes.lf) {
              chunks.push(codes.crlf)
              k++
            } else {
              chunks.push(codes.vcr)
            }

            column = 1
            break
          case code === codes.ht:
            /**
             * Next column.
             *
             * @const {number} n
             */
            const n: number = Math.ceil(column / tabSize) * tabSize

            chunks.push(codes.vht)
            while (column++ < n) chunks.push(codes.vs)

            break
          case code === codes.lf:
            chunks.push(codes.vlf)
            column = 1
            break
          default:
            chunks.push(code)
            column++
            break
        }

        index += k
      }
    }

    return end && chunks.push(codes.eof), chunks
  }
}

export default preprocess
