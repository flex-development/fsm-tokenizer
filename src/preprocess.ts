/**
 * @file preprocess
 * @module vfile-lexer/preprocess
 */

import { codes } from './enums'
import type { PreprocessOptions } from './interfaces'
import type {
  Code,
  Column,
  Encoding,
  FileLike,
  Preprocessor,
  Value
} from './types'

/**
 * Create a preprocessor to turn a value into character code chunks.
 *
 * @see {@linkcode PreprocessOptions}
 * @see {@linkcode Preprocessor}
 *
 * @param {PreprocessOptions | null | undefined} [options] - Configuration
 * @return {Preprocessor} Character code preprocessor
 */
function preprocess(
  this: void,
  options?: PreprocessOptions | null | undefined
): Preprocessor {
  const { tabSize = 2 } = options ?? {}
  return preprocessor

  /**
   * Turn `value` into character code chunks.
   *
   * @param {FileLike | Value | null | undefined} value - Value to preprocess
   * @param {Encoding | null | undefined} [encoding] - Character encoding to use
   * when value or its contents is {@linkcode Uint8Array}
   * @param {boolean | null | undefined} [end] - End of stream?
   * @return {Code[]} Character code chunks
   */
  function preprocessor(
    value: FileLike | Value | null | undefined,
    encoding?: Encoding | null | undefined,
    end?: boolean | null | undefined
  ): Code[] {
    /**
     * Character code chunks.
     *
     * @const {Code[]} chunks
     */
    const chunks: Code[] = []

    if (
      (typeof value === 'string' && value) ||
      (typeof value === 'object' && value)
    ) {
      value = typeof value === 'object' && 'value' in value
        ? value.value
        : value

      value = typeof value === 'string'
        ? value.toString()
        : new TextDecoder(encoding ?? undefined).decode(value)

      /**
       * Current column.
       *
       * @var {Column} column
       */
      let column: Column = 1

      /**
       * Index of current character code.
       *
       * @var {number} index
       */
      let index: number = 0

      while (index < value.length) {
        /**
         * Character code.
         *
         * @var {NonNullable<Code>} code
         */
        let code: NonNullable<Code> = value[index]!.codePointAt(0)!

        /**
         * Difference between next column and current column.
         *
         * @var {number} k
         */
        let k: number = 1

        switch (true) {
          case code === codes.cr:
            if (value[index + 1]?.codePointAt(0) === codes.lf) {
              chunks.push(codes.crlf)
              k++
            } else {
              chunks.push(codes.vcr)
            }

            column = 1
            break
          case code === codes.ht:
            /**
             * Next column.
             *
             * @const {number} n
             */
            const n: number = Math.ceil(column / tabSize) * tabSize

            chunks.push(codes.vht)
            while (column++ < n) chunks.push(codes.vs)

            break
          case code === codes.lf:
            chunks.push(codes.vlf)
            column = 1
            break
          default:
            chunks.push(code)
            column++
            break
        }

        index += k
      }
    }

    return end && chunks.push(codes.eof), chunks
  }
}

export default preprocess
