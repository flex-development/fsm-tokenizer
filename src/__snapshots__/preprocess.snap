// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:preprocess > preprocess > should return character code chunk list (0) 1`] = `
[
  -13,
  null,
]
`;

exports[`unit:preprocess > preprocess > should return character code chunk list (1) 1`] = `
[
  -2,
  -1,
  null,
]
`;

exports[`unit:preprocess > preprocess > should return character code chunk list (2) 1`] = `
[
  -4,
  -5,
  -3,
  null,
]
`;

exports[`unit:preprocess > preprocess > should return character code chunk list (3) 1`] = `
[
  104,
  101,
  108,
  108,
  111,
  -2,
  119,
  111,
  114,
  108,
  100,
  null,
]
`;

exports[`unit:preprocess > preprocess > should return character code chunk list (4) 1`] = `
[
  45,
  45,
  100,
  101,
  98,
  117,
  103,
  61,
  116,
  114,
  117,
  101,
  32,
  45,
  45,
  112,
  105,
  122,
  122,
  97,
  45,
  116,
  121,
  112,
  101,
  32,
  109,
  101,
  97,
  116,
  32,
  45,
  45,
  115,
  109,
  97,
  108,
  108,
  null,
]
`;
